### PR TITLE
Fixes #233 by implementing support for triggered demand in in the SubscriberBlackboxVerification

### LINKED ISSUE #233: Question on validity of SubscriberBlackboxVerification.required_spec201_blackbox_mustSignalDemandViaSubscriptionRequest
Hi folks,

This test is currently failing when run against Vert.x.

The test appears to expect Subscription.request(..) to be called sometime shortly after the onSubscribe method of the Subscriber is called.

From the docs, it appears this is trying to validate Subscriber requirement 2.1:

https://github.com/reactive-streams/reactive-streams-jvm#2.1

"A Subscriber MUST signal demand via Subscription.request(long n) to receive onNext signals."

However, I can't see anything in the above that says request() must be invoked when the onSubscribe method is called.

The Vert.x subscriber implementation won't call Subscription.request until some time later when the user has actually set a handler on the subscriber.

So I would like to question the validity of this test.

### PR DESCRIPTION
**Pull request by viktorklang**
Fixes #233 @purplefox & @ktoso?

### CODE REVIEW

File: tck/src/main/java/org/reactivestreams/tck/SubscriberBlackboxVerification.java [hunk 1]
```diff
@@ -48,6 +48,16 @@ protected SubscriberBlackboxVerification(TestEnvironment env) {
...
   public abstract Subscriber<T> createSubscriber();
 
+  /**
+   * Override this method if the Subscriber implementation you are verifying
+   * needs an external signal before it signals demand to its Publisher.
+   *
+   * By default this method does nothing.
+   */
+  public void triggerRequest(final Subscriber<? super T> subscriber) {
+
```
Comment (ktoso) [inline tck/src/main/java/org/reactivestreams/tck/SubscriberBlackboxVerification.java]: "this <strike>page</strike> method body was intentionally left blank." :wink:

File: tck/src/test/java/org/reactivestreams/tck/SyncTriggeredDemandSubscriberTest.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  @Override public Subscriber<Integer> createSubscriber() {
+    return new SyncTriggeredDemandSubscriber<Integer>() {
+      private long acc;
+      @Override protected long foreach(final Integer element) {
+        acc += element;
+        return 1;
+      }
+
+      @Override public void onComplete() {
+        System.out.println("Accumulated: " + acc);
```
Comment (ktoso) [inline tck/src/test/java/org/reactivestreams/tck/SyncTriggeredDemandSubscriberTest.java]: remove before merge?

File: tck/src/test/java/org/reactivestreams/tck/support/SyncTriggeredDemandSubscriber.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
+package org.reactivestreams.tck.support;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+/**
+ * SyncSubscriber is an implementation of Reactive Streams `Subscriber`,
+ * it runs synchronously (on the Publisher's thread) and requests one element
+ * at a time and invokes a user-defined method to process each element.
```
Comment (ktoso) [inline tck/src/test/java/org/reactivestreams/tck/support/SyncTriggeredDemandSubscriber.java]: Add a small notice about why/how it's a "triggered" one.

File: tck/src/main/java/org/reactivestreams/tck/SubscriberBlackboxVerification.java [hunk 4]
```diff
@@ -74,6 +84,7 @@ public void required_spec201_blackbox_mustSignalDemandViaSubscriptionRequest() t
     blackboxSubscriberTest(new BlackboxTestStageTestRun() {
       @Override
       public void run(BlackboxTestStage stage) throws InterruptedException {
+        triggerRequest(stage.subProxy().sub());
```
Comment (ktoso) [inline tck/src/main/java/org/reactivestreams/tck/SubscriberBlackboxVerification.java]: looks good

File: tck/src/main/java/org/reactivestreams/tck/SubscriberBlackboxVerification.java [hunk 5]
```diff
@@ -248,6 +259,7 @@ public void run(BlackboxTestStage stage) throws Throwable {
         final BlackboxSubscriberProxy<T> probe = stage.createBlackboxSubscriberProxy(env, sub);
 
         pub.subscribe(probe);
+        triggerRequest(sub);
```
Comment (ktoso) [inline tck/src/main/java/org/reactivestreams/tck/SubscriberBlackboxVerification.java]: :+1:

File: tck/src/test/java/org/reactivestreams/tck/support/SyncTriggeredDemandSubscriber.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        (new IllegalStateException(s + " violated the Reactive Streams rule 3.15 by throwing an exception from cancel.", t)).printStackTrace(System.err);
+      }
+    } else {
+      // We have to assign it locally before we use it, if we want to be a synchronous `Subscriber`
+      // Because according to rule 3.10, the Subscription is allowed to call `onNext` synchronously from within `request`
+      subscription = s;
+    }
+  }
+
+  public boolean triggerDemand(final long n) {
```
Comment (ktoso) [inline tck/src/test/java/org/reactivestreams/tck/support/SyncTriggeredDemandSubscriber.java]: add comments before merge here please

**Comment by ktoso** [comment]
This looks v. good I think, thanks Viktor!

File: tck/src/test/java/org/reactivestreams/tck/support/SyncTriggeredDemandSubscriber.java (hunk 3 above)
Comment (viktorklang) [inline tck/src/test/java/org/reactivestreams/tck/support/SyncTriggeredDemandSubscriber.java]: :+1:

File: tck/src/test/java/org/reactivestreams/tck/SyncTriggeredDemandSubscriberTest.java (hunk 2 above)
Comment (viktorklang) [inline tck/src/test/java/org/reactivestreams/tck/SyncTriggeredDemandSubscriberTest.java]: :+1:

File: tck/src/main/java/org/reactivestreams/tck/SubscriberBlackboxVerification.java (hunk 4 above)
Comment (viktorklang) [inline tck/src/main/java/org/reactivestreams/tck/SubscriberBlackboxVerification.java]: Thanks Kon!

File: tck/src/test/java/org/reactivestreams/tck/support/SyncTriggeredDemandSubscriber.java (hunk 6 above)
Comment (viktorklang) [inline tck/src/test/java/org/reactivestreams/tck/support/SyncTriggeredDemandSubscriber.java]: :+1:

**Comment by viktorklang** [comment]
@ktoso @purplefox & @reactive-streams/contributors Addressed review feedback in the following commit: https://github.com/reactive-streams/reactive-streams-jvm/commit/658c95eaed8d47de1076eeaac0abb7e12e767140

**Comment by viktorklang** [comment]
@reactive-streams/contributors & @purplefox This PR is ready to merge, Tim please verify that this fixes your issue and contributors-please raise any objections you might have (this is completely without impact on already TCK-passing implementations). I'll merge this on the 30th of March unless there are objections. Thanks

**Comment by purplefox** [comment]
This works for me.

(Aside: One small issue, not specifically related to this PR, I had to install Java 7 to build the TCK as it doesn't build using Java 8 - small JavaDoc issue. Java 7 will be EOL in a month)

**Comment by viktorklang** [comment]
Great that it worked. What was the Java Doc issue?

-- 
Cheers,
√

**Comment by viktorklang** [comment]
@purplefox Alright, I found and fixed the issue, it was due to the requirement to qualify nested classes in Javadoc links.

**Pull request merged**